File: packages/java/ygo74-agent-runtime/src/main/java/com/ygo74/agentruntime/routing/Dispatcher.java
package com.ygo74.agentruntime.routing;

import com.ygo74.agentruntime.domains.contracts.ExchangeModels.StandardExchangeRequest;
import com.ygo74.agentruntime.domains.contracts.ExchangeModels.StandardExchangeResponse;

import java.util.function.Function;

public interface Dispatcher {
    StandardExchangeResponse dispatch(StandardExchangeRequest request, Function<String, UseCaseHandler> resolver);

    interface UseCaseHandler {
        StandardExchangeResponse handle(StandardExchangeRequest request);
    }
}

File: packages/python/agents/ygo74/agent_runtime/routing/dispatcher.py
from collections.abc import Callable
from typing import Protocol

from ygo74.agent_runtime.domains.contracts.exchange_models import StandardExchangeRequest, StandardExchangeResponse


class UseCaseHandler(Protocol):
    def __call__(self, request: StandardExchangeRequest) -> StandardExchangeResponse:
        ...


class Dispatcher(Protocol):
    def dispatch(self, request: StandardExchangeRequest, resolver: Callable[[str], UseCaseHandler]) -> StandardExchangeResponse:
        ...
